### PR TITLE
feat(jax): add 1g1p intranode and internode deepep for jax

### PR DESCRIPTION
**Pull request by llying-001**
## Summary

This PR adds **JAX DeepEP internode (multi-node) support** plus a new **1-GPU-per-process launch mode** for both intranode and internode DeepEP dispatch/combine, on top of the existing in-process (pmap) path.

Previously the JAX DeepEP bindings only supported the in-process (single-process, multi-GPU via `pmap`) execution model and were intranode-only. To run DeepEP at MoE training scale on ROCm we need:

1. A **per-process** execution model so each rank owns exactly one GPU and synchronises via IPC handles + rocSHMEM (matching the torch `deep_ep.Buffer(group=ProcessGroup, ...)` topology).
2. **Internode** dispatch/combine kernels exposed to JAX so EP can span multiple nodes over RDMA.
3. A way to **pin the EP communication domain to a subset of JAX processes** (e.g. one mesh axis carries EP while others carry FSDP/DP).

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Infra/Build change
- [ ] Code refactoring

## Changes

- **New JAX DeepEP runtime layer** (`primus_turbo/jax/deep_ep/runtime.py`, `deep_ep/__init__.py`):
  - Introduces `LaunchMode` with `MODE_INPROC` and `MODE_PER_PROCESS`, selected via the `PRIMUS_TURBO_JAX_DEEPEP_MODE` env var.
  - Adds `set_ep_group(ep_ranks)` / `reset_runtime()` to decouple the EP rank set from the full JAX world; warns once when EP falls back to `jax.process_count()`.
  - `ensure_deepep_runtime()` bootstraps the C++ buffer once per process; IPC handles and the rocSHMEM unique id are exchanged via a JAX `kv-store` all-gather instead of `jax.lax.all_gather`.
- **New JAX primitives & lax wrappers for internode DeepEP** (`primitive/moe/moe_dispatch.py`, `primitive/moe/moe_combine.py`, `lax/moe/moe_dispatch_combine.py`):
  - `moe_internode_dispatch_p`, `moe_internode_cached_dispatch_p`, `moe_internode_combine_p` and their `per_process` FFI lowerings.
  - Existing `moe_dispatch` / `moe_combine` now thread `ep_size` and `launch_mode` attrs down to the FFI layer and validate them against the visible device count.
  - Adds a `warmup(hidden_bytes, ...)` helper to eagerly initialise the runtime outside `jax.jit`.
- **C++ side** (`csrc/jax/bindings_jax.cpp`, `csrc/jax/deep_ep/deep_ep.{h,cpp}`, `csrc/jax/deep_ep/handler.cpp`, `csrc/jax/extensions.h`):
  - New FFI handlers: `moe_dispatch_inproc` / `moe_dispatch_per_process` (and cached/combine variants) plus `moe_internode_{dispatch,cached_dispatch,combine}_per_process`.
  - New `Buffer::InternodeDispatch` / `Buffer::InternodeCombine` plumbing wired to the existing rocSHMEM kernels.
  - New pybind submodule `primus_turbo.jax._C.deep_ep` exposing `create_per_process_buffer`, `sync_per_process_buffer`, `destroy_per_process_buffer`, `is_per_process_buffer_ready`, `per_process_buffer_nvl_bytes`, and (when rocSHMEM is enabled) `get_unique_id`.
  - `Buffer` lifecycle hardening: explicit `Destroy()` with safe fallback in the destructor, NVL peer barrier before closing IPC handles, IPC-handle close only when `ipc_synced_`, rocSHMEM finalize guarded by `DISABLE_ROCSHMEM`, plus null-pointer checks.
- **Bug fixes**:
  - Fix combine hang caused by wrong handle-matrix order in cached dispatch path.
  - Remove the device-side spin on `moe_recv_counter_mapped` in `notify_dispatch` when `num_worst_tokens > 0` (XLA static-shape path) — the host skips its CPU sync there, so back-to-back launches would otherwise spin forever on a stale counter (`csrc/kernels/deep_ep/internode.cu`).
  - Fix compile error introduced during the internode rewiring.
- **Build** (`tools/build_utils.py`):
  - Auto-discover linkable system libraries (e.g. `libionic`) via `LD_LIBRARY_PATH`/`LIBRARY_PATH`/multiarch defaults when wiring `librocshmem.a`.
- **Tests & infra**:
  - New `tests/jax/lax/test_mp_dispatch_combine.py` covering multi-process (1 GPU/proc) dispatch/combine via the `per_process` mode, with a new `JaxMultiProcessTestCase` helper in `tests/jax/test_utils.py`.
  - `tests/conftest.py` recognises `JaxMultiProcessTestCase` under `--dist-only`.
  - Remove the obsolete internode UT test.
- **Docs**: updated `benchmark/README.md` with the multi-node DeepEP launch instructions.

# Checklist:

- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/jax/test_utils.py [hunk 1]
```diff
@@ -104,3 +112,119 @@ def assert_allclose(actual, expected, rtol=1e-5, atol=1e-8):
...
+# Multi-process test harness (1 GPU per process)
+###################################################################
+
+
+def _find_free_port():
+    """Return an ephemeral port that is (momentarily) free on localhost."""
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        s.bind(("", 0))
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        return s.getsockname()[1]
```
[the remark this answers is not in the document]
Comment (llying-001) [reply tests/jax/test_utils.py:135]: Acknowledged. _find_free_port is racy by construction (closed-port handoff), but the practical hit rate on our CI is negligible. Tracking as a follow-up to add a retry-on-EADDRINUSE wrapper around jax.distributed.initialize in a separate hardening PR so this one stays scoped to the 1g1p / internode feature.

File: primus_turbo/jax/lax/moe/moe_dispatch_combine.py [hunk 2]
```diff
@@ -384,6 +869,55 @@ def _moe_combine_impl(
         rank_prefix_matrix,
         channel_prefix_matrix,
         send_head,
```
[the remark this answers is not in the document]
Comment (llying-001) [reply primus_turbo/jax/lax/moe/moe_dispatch_combine.py:915]: erified against the pytorch backend: the unpack rank_prefix_matrix, _, channel_prefix_matrix, src_idx, is_recv_token_in_rank, send_head = handle matches Buffer.combine in pytorch/deep_ep/buffer.py:577 exactly — handle[1] (the send-side matrix) is intentionally discarded with _, and handle[2] (the dispatch-time recv_channel_prefix_matrix) is bound to the local name channel_prefix_matrix to match the C++ kernel's parameter name. cached_dispatch uses handle[1] because it re-runs dispatch (send side); combine starts from the dispatch recv side and therefore needs the recv-side layout. No change needed.

File: primus_turbo/jax/deep_ep/runtime.py [hunk 3]
```diff
@@ -0,0 +1,698 @@
...
+
+    _bootstrap_barrier_call_id += 1
+    label = f"primus_turbo.deep_ep.bootstrap.cid={_bootstrap_barrier_call_id}"
+    try:
+        sync_global_devices(label)
+    except Exception:  # pragma: no cover  pylint: disable=broad-except
+        # A failed barrier is strictly worse than no barrier — propagate so
+        # the caller's existing exception path destroys partial C++ state.
+        log.exception("DeepEP bootstrap barrier %r failed", label)
+        raise
```
[the remark this answers is not in the document]
Comment (llying-001) [reply primus_turbo/jax/deep_ep/runtime.py:525]: In actual training, there are multiple EP groups, and all of them will reach this synchronization point during initialization.

**Review by zhenhuang12 — APPROVED** [review]
LGMT

**Review by xiaobochen-amd — APPROVED** [review]
